src/sql.rs

```diff
@@ -299,10 +303,25 @@ impl Sql {
         }
     }
 
-    /// Allocates a connection and calls given function with the connection.
+    /// Locks the write transactions mutex. We use
+    /// [DEFERRED transactions](https://www.sqlite.org/lang_transaction.html#deferred_immediate_and_exclusive_transactions)
+    /// for more parallelism; therefore we need to serialize db writers, otherwise a write
```

Comment:
Yes, we can make all write transactions IMMEDIATE so that they start as write, not read ones, and upgrading them later is not needed. But we anyway should make read transactions DEFERRED to allow them run in parallel. And i think there's not much complexity in upgrading a transaction to write -- with this PR SQLite will just check that there is no another write performed in parallel and mark the transaction as write.

Reply:
Rewrote the comment completely to make it more clear